### PR TITLE
[Feature] Running end-to-end tests on local machine

### LINKED ISSUE #590: [Feature] Running end-to-end tests on local machine
### Search before asking

- [X] I had searched in the [issues](https://github.com/ray-project/kuberay/issues) and found no similar feature requirement.

### Description

We need to download (`docker pull`) all necessary docker images before we load the images into a KinD cluster. Without this, the command `kind load docker-image kuberay/apiserver:nightly` will report an error message as follows.

```
ERROR: image: "kuberay/apiserver:nightly" not present locally
```
 
In addition, we also use `kind load` command to preload the `rayproject/ray` image into KinD cluster. Although the Kubernetes cluster will pull the ray image automatically when we create a custom resource `RayCluster`, pulling `rayproject/ray:1.9.0` (2.44GB) may cause `ImagePullBackOff` ([Link](https://github.com/kubernetes/kubernetes/issues/59376)) due to timeout. In addition, the `RayCluster` startup time will vary based on network speed and make the tests flaky.

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by kevin85421**
## Why are these changes needed?
We need to download (`docker pull`) all necessary docker images before we load the images into a KinD cluster. Without this, the command `kind load docker-image kuberay/apiserver:nightly` will report an error message as follows.

```
ERROR: image: "kuberay/apiserver:nightly" not present locally
```
 
In addition, we also use `kind load` command to preload the `rayproject/ray` image into KinD cluster. Although the Kubernetes cluster will pull the ray image automatically when we create a custom resource `RayCluster`, pulling `rayproject/ray:1.9.0` (2.44GB) may cause `ImagePullBackOff` ([Link](https://github.com/kubernetes/kubernetes/issues/59376)) due to timeout. In addition, the `RayCluster` startup time will vary based on network speed and make the tests flaky.

## Related issue number
Closes #590 

## Checks

```bash
cd /path/to/your/kuberay
python3 tests/compatibility-test.py
```
截圖 2022-09-26 上午11 10 07

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: tests/compatibility-test.py [hunk 1]
```diff
@@ -54,10 +58,11 @@ def create_cluster():
 
 
 def apply_kuberay_resources():
-    shell_assert_success('kind load docker-image kuberay/operator:{}'.format(kuberay_sha))
-    shell_assert_success('kind load docker-image kuberay/apiserver:{}'.format(kuberay_sha))
-    shell_assert_success(
-        'kubectl create -k manifests/cluster-scope-resources')
+    shell_assert_success('kind load docker-image {}'.format(kuberay_operator_image))
```
Comment (wilsonwang371) [inline tests/compatibility-test.py:61]: we need to push the correct build to kind cluster. not the nightly build.

Comment (kevin85421) [reply tests/compatibility-test.py]: The variable `kuberay_operator_image` will be updated in the function `parse_environment()`. Hence, the value of `kuberay_operator_image` is correct.

The previous CI errors are caused by `docker pull`. The new image builds (e.g. `kuberay/operator:1ebf6a8`) only exist in the local docker registry, and thus errors will be reported when we try to pull `kuberay/operator:1ebf6a8` from DockerHub. Hence, I updated the function `download_images` to pull images only when the images do not exist in the local docker registry.

**Comment by wilsonwang371** [comment]
I just realized we may need to separate these functions into different files. that will be easier for our future updates. I posted another patch https://github.com/ray-project/kuberay/pull/591 and this should  make our changes to ci tests more convenient.

**Comment by kevin85421** [comment]
> I just realized we may need to separate these functions into different files. that will be easier for our future updates. I posted another patch #591 and this should make our changes to ci tests more convenient.

Thank you for this patch! Feel free to ping me for review when it passes CI/CD tests.

File: ray-operator/DEVELOPMENT.md [hunk 2]
```diff
@@ -116,4 +116,12 @@ make sync
 
 # Reproduce CI error for job "helm-chart-verify-rbac" (consistency 5)
 python3 ../scripts/rbac-check.py
-```
\ No newline at end of file
+```
+
+### Run end-to-end tests locally
+
+We have some [end-to-end tests](https://github.com/ray-project/kuberay/blob/master/.github/workflows/actions/compatibility/action.yaml) on GitHub Actions.
```
Comment (DmitriGekhtman) [inline ray-operator/DEVELOPMENT.md:123]: I'd add a couple of sentences here:

`These tests operate small Ray clusters running within a [kind](https://kind.sigs.k8s.io/) (Kubernetes-in-docker) environment. To run the tests yourself, follow these steps:`

Comment (DmitriGekhtman) [reply ray-operator/DEVELOPMENT.md:123]: Hmm, looks like running a test requires specifying a KubeRay operator SHA from a commit that's already merged to master?

How would you test your local development operator code? Maybe we can make the full operator image configurable via environment variable?

Comment (kevin85421) [reply ray-operator/DEVELOPMENT.md:123]: Good idea. I will update it.

**Comment by kevin85421** [comment]
I found an interesting bug when I tried to use [argparse](https://docs.python.org/3/library/argparse.html) to simplify the logic of `parse_environments()` (See https://github.com/ray-project/kuberay/pull/589/commits/6751bc93c1dce70d0d971d8f9755c95ca325c99c for more details.). The following simple scripts can reproduce the bugs.

* `test1.py`
  ```python
   import argparse
   import unittest
  
   if __name__ == '__main__':
       parser = argparse.ArgumentParser()
       parser.add_argument('--input', '-i', dest='input',
                           required=False, default='Hello KubeRay')
       args = parser.parse_args()
       print(args.input)
       unittest.main()
  ```

* `test2.py`
  ```python
   import argparse
   import unittest
  
   if __name__ == '__main__':
       parser = argparse.ArgumentParser()
       parser.add_argument('--input', '-i', dest='input',
                           required=False, default='Hello KubeRay')
       args = parser.parse_args()
       print(args.input)
  ```

* Experiment1: `python3 test1.py --input haha` 
   ```bash
  $ python3 test.py --input haha
  haha
  usage: test.py [-h] [-v] [-q] [--locals] [-f] [-c] [-b] [-k TESTNAMEPATTERNS]
                 [tests [tests ...]]
  test.py: error: unrecognized arguments: --input
  (base)
   ```

* Experiment2: `python3 test2.py --input haha` 
  ```bash
  $ python3 test.py --input haha
  haha
  (base)
  ```

The error message "unrecognized arguments: --input" in Experiment1 is reported by `unittest.main()`. This is because `unittest.main()` implicitly read `sys.argv` (See [Link](https://docs.python.org/3/library/unittest.html#unittest.main) for more details.)

File: ray-operator/DEVELOPMENT.md (hunk 2 above)
Comment (kevin85421) [reply ray-operator/DEVELOPMENT.md:123]: Updated.

**Review by DmitriGekhtman — APPROVED** [review]
Thanks, this is a great improvement.

**Pull request merged**